File: cpp-ami/Client-callback.cpp
#include <Ice/Ice.h>
#include <factorial.h>

using namespace std;
using namespace Ice;
using namespace Example;

class FactorialCB : public IceUtil::Shared {
public:
  void response(const Ice::Long retval) {
    cout << "Callback: Value is " << retval << endl;
  }

  void failure(const Exception& ex) {
    cout << "Exception is: " << ex << endl;
  }
};

class Client: public Ice::Application {
public:
  int run(int argc, char* argv[]) {

    ObjectPrx proxy = communicator()->stringToProxy(argv[1]);
    MathPrx math = MathPrx::checkedCast(proxy);

    Callback_Math_factorialPtr factorial_cb =
      newCallback_Math_factorial(new FactorialCB,
				 &FactorialCB::response,
				 &FactorialCB::failure);

    math->begin_factorial(atoi(argv[2]), factorial_cb);
    return 0;
  }

};

int main(int argc, char* argv[]) {
  if (argc != 3) {
    cerr << "usage: " << argv[0] << "<server> <value>" << endl;
    return 1;
  }

  Client app;
  return app.main(argc, argv);
}

File: java-ami/Client_callback.java
class FactorialCB extends Example.Callback_Math_factorial {
  public void response(long value) {
    System.out.println("Callback: Value is: " + value);
  }

  public void exception(Ice.LocalException ex) {
    System.err.println("Exception is: " + ex);
  }
}

public class Client_callback extends Ice.Application {
  public int run(String[] args) {
    Ice.ObjectPrx proxy = communicator().stringToProxy(args[0]);
    Example.MathPrx math = Example.MathPrxHelper.checkedCast(proxy);

    FactorialCB factorial_cb = new FactorialCB();
    math.begin_factorial(Integer.parseInt(args[1]), factorial_cb);
    return 0;
  }

  static public void main(String[] args) {
    if (args.length != 2) {
      System.err.println(appName() + ": usage: <server> <value>");
      return;
    }

    Client_callback app = new Client_callback();
    app.main("Client", args);
  }
}
